pkg/dns-registrar/requester/requester.go

```diff
@@ -100,18 +124,29 @@ func NewDoHRequester(dohurl string, domain string, pubkey []byte, utlsDistributi
 	}, nil
 }
 
-// New Requester using plain UDP as transport
+// New Requester using UDP as transport with default dialer
 func NewUDPRequester(remoteAddr net.Addr, domain string, pubkey []byte) (*Requester, error) {
+	listener := net.ListenConfig{}
+	return NewUDPRequesterWithListenPacket(remoteAddr, domain, pubkey, listener.ListenPacket)
+}
+
+// New Requester using UDP as transport
+func NewUDPRequesterWithListenPacket(remoteAddr net.Addr, domain string, pubkey []byte, listenPacket func(ctx context.Context, network, address string) (net.PacketConn, error)) (*Requester, error) {
```

Comment:
The issue is that `UDPConn.WriteTo` that implements `net.PacketConn` does not work when it is "dialed" and seems to only work if it is "listened". I did try using the dialer with `"udp"` as network type and asserting it to a `net.UDPConn`, and there seems to be some internal state in `UDPConn` that keeps track of whether it was called via `dial` and returns an error on every call to `WriteTo`. `ListenPacket` was chosen because it has the closest function signature that I could find.

Reply:
okay I was playing with this in an example program and the error that I was getting is something like `use of writeto with pre-connected connection` when trying to use a dialed `UDPConn` as a `PacketConn`. I haven't found a good way around this unfortunately. If we have to have a `listenPacket` function we should ensure a way for the downstream library to provide that function to the `DNSRegistrar` (either through the constructor or through the dialer/session/etc.). We might also want to add a `type ListenerContextFunc func(...)` so we are not passing around the function signature. 

It is a bummer that golang networking has to be like this.

Reply:
One alternative would be to modify the `DNSRegistrar` to use a `UDPConn` instead of a `PacketConn`. That would allow us to use the same dial function signature, but we would need to modify the `recvLoop`  and `sendLoop` functions to use `Read` and `Write`.

Alternatively we could add a second DNSRegistrar type that functions basically the same and has a different constructor so we could build using either `PacketConn` or `UDPConn`. 

Just providing some options here. lmk what you think sounds easier / cleaner.

Reply:
I've now made the DNS registrar take a `Conn` instead of `PacketConn`, which lets us use our dialer with UDP. While doing it, I added some stub functions in `QueuePacketConn` to make it implement the `Conn` interface that the `DNSPacketConn` now expects. 

I think the reason I chose `PacketConn` initially is that the connectionless nature of `PacketConn` is similar to DNS, but for our use case, we just need to connect to one remote so a `Conn` is better. Therefore we can actually remove the `QueuePacketConn` dependency altogether, but I think we can do that in another PR as there shouldn't be any API changes.